File: softeer/나무수확.java
import java.util.*;
import java.io.*;

class Main{
    static int n;
    static int map[][];
    static int dp[][];
    static int rdp[][];

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        n = Integer.parseInt(br.readLine());

        map = new int[n+2][n+2];
        dp = new int[n+2][n+2];
        rdp = new int[n+2][n+2];

        for(int i = 1; i <= n; i++){
            String inp[] = br.readLine().split(" ");
            for(int j = 1; j <= n; j++){
                map[i][j] = Integer.parseInt(inp[j-1]);
            }
        }

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= n; j++){
                dp[i][j] = Math.max(dp[i-1][j],dp[i][j-1]) + map[i][j];
                rdp[n+1-i][n+1-j] = Math.max(rdp[n+2-i][n+1-j],rdp[n+1-i][n+2-j]) + map[n+1-i][n+1-j]; 
            }
        }

        // for(int i = 1; i <= n; i++){
        //     for(int j = 1; j <= n; j++){
        //         bw.append(dp[i][j]  + " ");
        //     }
        //     bw.append("\n");
        // }
        // bw.append("---------------\n");
        // bw.flush();
        
        // for(int i = 1; i <= n; i++){
        //     for(int j = 1; j <= n; j++){
        //         bw.append(rdp[i][j]  + " ");
        //     }
        //     bw.append("\n");
        // }

        int maxVal = 0;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= n; j++){
                if(maxVal < dp[i][j] + rdp[i][j]){
                    maxVal = dp[i][j] + rdp[i][j];
                }
            }
        }

        bw.append(maxVal + "\n");
        bw.flush();
        br.close();
        bw.close();
    }
}

File: softeer/나무수확.cpp
#include <iostream>

using namespace std;

int n,answer = 0;
int map[1001][1001];
int dp[1001][1001];
int rdp[1002][1002];

int main(){
    cin >> n;

    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++){
            cin >> map[i][j];
            dp[i][j] = max(dp[i-1][j],dp[i][j-1]) + map[i][j];
        }
    }  

    for(int i = n; i > 0; i--){
        for(int j = n; j > 0; j--){
            rdp[i][j] = max(rdp[i+1][j],rdp[i][j+1]) + map[i][j];
        }
    }
    int col = 0, row = 0, maxVal = 0;
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++){
            if(maxVal < rdp[i][j] + dp[i][j]){
                maxVal = rdp[i][j] + dp[i][j];
                row = i;
                col = j;
            }
        }
    }

    // for(int i = 1; i <= n; i++){
    //     for(int j = 1; j <= n; j++){
    //         cout << dp[i][j] << " ";
    //     }
    //     cout << "\n";
    // }

    // cout << "------------------\n";

    // for(int i = 1; i <= n; i++){
    //     for(int j = 1; j <= n; j++){
    //         cout << rdp[i][j] << " ";
    //     }
    //     cout << "\n";
    // }

    answer = dp[row][col] + rdp[row][col];

    cout << answer << "\n";
    return 0;
}